broker.go

```diff
@@ -0,0 +1,706 @@
+	// consumer. By default set to 2000000 bytes.
+	MaxFetchSize int32
+
+	// Consumer cursor starting point. Set to StartOffsetNewest to receive only
+	// newly created messages or StartOffsetOldest to read everything. Assign
+	// any offset value to manually set cursor. Defaults to StartOffsetOldest
+	StartOffset int64
+
+	// Logger used by consumer. By default, reuse logger assigned to broker.
+	Log Logger
+}
+
+// NewConsumerConf return default consumer configuration.
+func NewConsumerConf(topic string, partition int32) ConsumerConf {
+	return ConsumerConf{
+		Topic:          topic,
+		Partition:      partition,
+		RequestTimeout: 0,
+		RetryLimit:     -1,
+		RetryWait:      time.Millisecond * 250,
+		RetryErrLimit:  10,
+		RetryErrWait:   time.Millisecond * 500,
+		MinFetchSize:   1,
+		MaxFetchSize:   2000000,
+		StartOffset:    StartOffsetOldest,
+		Log:            nil,
+	}
+}
+
+// Consumer is representing single partition reading buffer. Consumer is also
+// providing limited failures handling and message filtering.
+type Consumer struct {
+	broker *Broker
+	conn   *connection
+	conf   ConsumerConf
+	offset int64
+	msgbuf []*proto.Message
+}
+
+// Consumer creates new consumer instance, bound to broker.
+func (b *Broker) Consumer(conf ConsumerConf) (consumer *Consumer, err error) {
+	conn, err := b.leaderConnection(conf.Topic, conf.Partition)
+	if err != nil {
+		return nil, err
+	}
+	if conf.Log == nil {
+		conf.Log = b.conf.Log
+	}
+	offset := conf.StartOffset
+	if conf.StartOffset < 0 {
+		switch conf.StartOffset {
+		case StartOffsetNewest:
+			off, err := b.OffsetLatest(conf.Topic, conf.Partition)
+			if err != nil {
+				return nil, err
+			}
+			// latest offset that next produced message will get, but consumer
+			// offset should be set to the last returned message offset
+			offset = off - 1
+		case StartOffsetOldest:
+			off, err := b.OffsetEarliest(conf.Topic, conf.Partition)
+			if err != nil {
+				return nil, err
+			}
+			offset = off
+			// this is stange kafka thing: asking about message with offset one
+			// higher than the last existing message is fine, except when there
+			// are no messages in the log -- in such case you should always ask
+			// about offset 0.
+			// Set offset -1 to make fetch request 0 offset message when there
+			// are none.
+			if off == 0 {
+				offset = -1
+			}
+		default:
+			return nil, fmt.Errorf("invalid start offset: %d", conf.StartOffset)
+		}
+	}
+	consumer = &Consumer{
+		broker: b,
+		conn:   conn,
+		conf:   conf,
+		msgbuf: make([]*proto.Message, 0),
+		offset: offset,
+	}
+	return consumer, nil
+}
+
+// Conf returns consumer configuration.
+func (c *Consumer) Conf() ConsumerConf {
+	return c.conf
+}
+
+// Fetch is returning single message from consumed partition. Consumer can
+// retry fetching messages even if responses return no new data. Retry
+// behaviour can be configured through RetryLimit and RetryWait consumer
+// parameters.
+//
+// Fetch is not thread safe.
```

Comment:
Shouldn't it be thread safe. I am not too sure there will not be a case where we want to fetch from multiple goroutines.